Mathlib/NumberTheory/ModularForms/JacobiTheta/TwoVariable.lean

```diff
@@ -0,0 +1,136 @@
+  · exact continuous_im.isOpen_preimage _ isOpen_Ioi
+  · exact fun i ξ (hξ : T < im ξ) ↦ jacobiTheta₂_term_bound hT.1 (le_refl _) hξ.le i
+
+/-- The two-variable Jacobi theta function is periodic in `τ` with period 2. -/
+lemma jacobiTheta₂_add_right (z τ : ℂ) : jacobiTheta₂ z (τ + 2) = jacobiTheta₂ z τ := by
+  refine tsum_congr (fun n ↦ ?_)
+  simp_rw [Complex.exp_add]
+  suffices cexp (π * I * n ^ 2 * 2 : ℂ) = 1 by rw [mul_add, Complex.exp_add, this, mul_one]
+  rw [(by push_cast; ring : (π * I * n ^ 2 * 2 : ℂ) = (n ^ 2 :) * (2 * π * I)), exp_int_mul,
+    exp_two_pi_mul_I, one_zpow]
+
+/-- The two-variable Jacobi theta function is periodic in `z` with period 1. -/
+lemma jacobiTheta₂_add_left (z τ : ℂ) : jacobiTheta₂ (z + 1) τ = jacobiTheta₂ z τ := by
+  refine tsum_congr (fun n ↦ ?_)
+  simp_rw [mul_add, mul_one, Complex.exp_add]
+  suffices cexp (2 * ↑π * I * n) = 1 by rw [this, mul_one]
+  simpa only [mul_comm] using exp_int_mul_two_pi_mul_I n
+
+/-- The two-variable Jacobi theta function is quasi-periodic in `z` with period `τ`. -/
+lemma jacobiTheta₂_add_left' (z τ : ℂ) :
+    jacobiTheta₂ (z + τ) τ = cexp (-π * I * (τ + 2 * z)) * jacobiTheta₂ z τ := by
+  conv_rhs => erw [← tsum_mul_left, ← (Equiv.addRight (1 : ℤ)).tsum_eq]
+  refine tsum_congr (fun n ↦ ?_)
+  rw [← Complex.exp_add, Equiv.coe_addRight, Int.cast_add]
+  ring_nf
+
+/-- The two-variable Jacobi theta function is even in `z`. -/
+lemma jacobiTheta₂_neg_left (z τ : ℂ) : jacobiTheta₂ (-z) τ = jacobiTheta₂ z τ := by
+  conv_lhs => rw [jacobiTheta₂, ← Equiv.tsum_eq (Equiv.neg ℤ)]
+  refine tsum_congr (fun n ↦ ?_)
+  rw [Equiv.neg_apply, Int.cast_neg, neg_sq]
+  ring_nf
+
+/-- The functional equation for the Jacobi theta function: `jacobiTheta₂ x τ` is an explict factor
+times `jacobiTheta₂ (x / τ) (-1 / τ)`. This is the key lemma behind the proof of the functional
+equation for Dirichlet L-series. -/
+theorem jacobiTheta₂_functional_equation (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    jacobiTheta₂ z τ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (-π * I * z ^ 2 / τ)
+    * jacobiTheta₂ (z / τ) (-1 / τ) := by
+  have h0 : τ ≠ 0; contrapose! hτ; rw [hτ, zero_im]
+  have h2 : 0 < (-I * τ).re
+  · simpa only [neg_mul, neg_re, mul_re, I_re, zero_mul, I_im, one_mul, zero_sub, neg_neg] using hτ
+  calc
+    _ = ∑' n : ℤ, cexp (-π * (-I * τ) * ↑n ^ 2 + 2 * π * (I * z) * ↑n) :=
+      tsum_congr (fun n ↦ by ring_nf)
+    _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * ∑' (n : ℤ), cexp (-π / (-I * τ) * (n + I * (I * z)) ^ 2) := by
+      rw [tsum_exp_neg_quadratic h2]
+    _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (π * I * (-1 / τ) * z ^ 2) * ∑' (n : ℤ),
+        cexp (2 * π * I * n * (z / τ) + π * I * n ^ 2 * (-1 / τ)) := by
+      simp_rw [mul_assoc _ (cexp _), ← tsum_mul_left (a := cexp _), ← Complex.exp_add]
+      congr 2 with n : 1; congr 1
+      field_simp [I_ne_zero]
+      ring_nf
+      rw [(by ring : I ^ 4 = (I ^ 2) ^ 2), I_sq]
```

Comment:
Maybe we should add `I^4=1`, since we will inevitably need it in the future.

Reply:
OK, I moved this to `Data.Complex.Basic`, but now I need to wait for almost the whole of mathlib to recompile! Time  for a lunch break 🥪 🍏 🫖